README.md
# api

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar api-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-06-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-06-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/api/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/api/compare/0.1.0...0.1.1


test/api/core_test.clj
(ns api.core-test
  (:require [api.core :refer :all]
            [clojure.test :refer :all]
            [clojure.spec.alpha :as s]
            [clojure.spec.gen.alpha :refer [generate]]))

(defn some-task-event [& [overwrites]]
  (into (generate (s/gen :app/task-event)) overwrites))

(deftest apply-event-test
  (testing "applying unknown event returns an error"
    (let [event {:event/name "some event"}]
      (is (thrown? clojure.lang.ExceptionInfo (apply-event [] event)))))

  (testing "`task-added` event is applied"
    (is (seq (apply-event [] (some-task-event {:event/name "task-added"})))))

  (testing "many `task-added` events with same uri are applied"
    (let [tasks (->> [(some-task-event
                        {:event/name "task-added"
                         :task/title "first event" :task/uri "uri-1"})
                      (some-task-event
                        {:event/name "task-added"
                         :task/title "second event" :task/uri "uri-1"})
                      (some-task-event
                        {:event/name "task-added"
                         :task/title "last event" :task/uri "uri-1"})]
                     (reduce apply-event []))]
      (is (= 1 (count tasks)))
      (is (= "last event" (:task/title (first tasks))))))

  (testing "`task-completed` event is applied"
    (is (empty? (->> (some-task-event {:event/name "task-completed"})
                     (apply-event []))))

    (is (empty? (->> [(some-task-event {:event/name "task-added"
                                        :task/uri   "1"})
                      (some-task-event {:event/name "task-completed"
                                        :task/uri   "1"})]
                     (reduce apply-event []))))

    (is (= 1 (->> [(some-task-event {:event/name "task-added"
                                     :task/uri   "1"})
                   (some-task-event {:event/name "task-completed"
                                     :task/uri   "2"})]
                  (reduce apply-event [])
                  (count))))

    (is (= 0 (->> [(some-task-event {:event/name "task-added"
                                     :task/uri   "1"})
                   (some-task-event {:event/name "task-completed"
                                     :task/uri   "1"})]
                  (reduce apply-event [])
                  (count))))))

(deftest in-memory-event-store-test
  (testing "it should keep a record of events"
    (let [{:keys [send, events]} (in-memory-event-store)]
      (is (empty? (events)))
      (send {:event/name "first event"})
      (is (= 1 (count (events))))
      (send {:event/name "another event"})
      (is (events) [{:event/name "first event"}
                    {:event/name "another event"}])))

  (testing "it should not store invalid events"
    (let [{:keys [send, events]} (in-memory-event-store)]
      (send {:some-prop "first event"})
      (is (= 0 (count (events))))))

  (testing "it should not share state with other stores"
    (is (empty? ((:events (in-memory-event-store)))))))

test/api/router_test.clj
(ns api.router-test
  (:require [api.core :refer :all]
            [clojure.test :refer :all]
            [muuntaja.core :as m]))

(defn parse-body [request] (m/decode "application/json" (:body request)))
(deftest http-routes-test
  (testing "it should respond to GET `/tasks`"
    (let [response ((app (in-memory-event-store))
                    {:request-method :get :uri "/tasks"})]
      (is (= 200 (:status response)))
      (is (= [] (parse-body response))))

    (let [bad-event-store (in-memory-event-store)]
      ((:send bad-event-store) {:event/name "bad event"})
      (is (= 500 (:status ((app bad-event-store)
                           {:request-method :get :uri "/tasks"}))))))

  (testing "it should respond to POST `/tasks/added`"
    (is (= 400 (-> {:request-method :post :uri "/tasks/added"}
                   ((app (in-memory-event-store)))
                   (:status))))

    (is (= 201 (-> {:request-method :post
                    :uri            "/tasks/added"
                    :body-params    {:uri "uri-1" :title "new task"}}
                   ((app (in-memory-event-store)))
                   (:status)))))

  (testing "it should respond to POST `/tasks/completed`"
    (is (= 400 (-> {:request-method :post, :uri "/tasks/completed"}
                   ((app (in-memory-event-store)))
                   (:status))))

    (is (= 201 (-> {:request-method :post
                    :uri            "/tasks/completed"
                    :body-params    {:uri "uri-1" :title "new task"}}
                   ((app (in-memory-event-store)))
                   (:status)))))

  (testing "it should return 404 when request uri unknown"
    (is (= 404 (-> {:request-method :get :uri "some-uri"}
                   ((app (in-memory-event-store)))
                   (:status)))))

  (testing "it should return 405 when request method is not allowed"
    (is (= 405 (-> {:request-method :post :uri "/tasks"}
                   ((app (in-memory-event-store)))
                   (:status))))))

test/api/features_test.clj
(ns api.features-test
  (:require [api.core :refer :all]
            [clojure.test :refer :all]
            [muuntaja.core :as m]))

(defn parse-body [request] (m/decode "application/json" (:body request)))
(deftest features-test
  (testing "it should add tasks"
    (let [app (app (in-memory-event-store))]
      (app {:request-method :post
            :uri            "/tasks/added"
            :body-params    {:uri "uri-1" :title "new task"}})
      (app {:request-method :post
            :uri            "/tasks/added"
            :body-params    {:uri "uri-2" :title "second task"}})
      (is (= 2 (count (parse-body (app {:request-method :get :uri "/tasks"})))))))

  (testing "it should complete tasks"
    (let [app (app (in-memory-event-store))]
      (app {:request-method :post
            :uri            "/tasks/added"
            :body-params    {:uri "uri-1" :title "new task"}})
      (is (= 1 (count (parse-body (app {:request-method :get :uri "/tasks"})))))
      (app {:request-method :post
            :uri            "/tasks/completed"
            :body-params    {:uri "uri-1" :title "new task"}})
      (is (= 0 (count (parse-body (app {:request-method :get :uri "/tasks"}))))))))

src/api/core.clj
(ns api.core
  (:require [clojure.spec.alpha :as s]
            [reitit.ring :as r]
            [reitit.ring.coercion :as rrc]
            [reitit.coercion.spec]
            [muuntaja.middleware :as mw]
            [ring.adapter.jetty :as j])
  (:gen-class))

(s/def :task/uri (s/and string? not-empty))
(s/def :task/title (s/and string? not-empty))
(s/def ::task (s/keys :req [:task/uri :task/title]))

(s/def :event/name (s/and string? not-empty))
(s/def ::event (s/keys :req [:event/name]))
(s/def :app/task-event (s/merge ::event ::task))

(s/def :error/reason (s/and string? not-empty))
(s/def :error/fault (s/and string? not-empty))
(s/def :error/details any?)
(s/def :app/system-error (s/keys :req [:error/reason :error/fault :error/details]))

;;; --------------------------- state reducer --------------------------------

(defn apply-event
  "Reduces a valid application state from a vector of task events. The
  events must be sorted in order of when the events were received."
  [tasks event]
  (case (:event/name event)
    "task-added"
    (->> (select-keys event [:task/uri :task/title])
         (conj (remove #(= (:task/uri event) (:task/uri %)) tasks)))
    "task-completed"
    (remove #(= (:task/uri %) (:task/uri event)) tasks)
    ;;; else
    (throw (ex-info "cannot apply event" event))))

;;; --------------------------- HTTP server ----------------------------------

(defn handle-task-event
  "Sends the received event to the event-store and responds to the request."
  [event-name send-to-store]
  (fn [request]
    (let [event {:event/name event-name
                 :task/uri   (get-in request [:body-params :uri])
                 :task/title (get-in request [:body-params :title])}]
      (when (s/valid? :app/task-event event) (send-to-store event)))
    {:status 201}))

(defn tasks-snapshot-from [events-from-store]
  "Returns a snapshot of tasks from the history of task events. Only responds
   when the snapshot was created successfully, otherwise returns an error."
  (fn [_]
    (try
      {:status 200 :body (reduce apply-event [] (events-from-store))}
      (catch clojure.lang.ExceptionInfo _ {:status 500}))))

(defn app
  "Listens to HTTP requests and routes them to the correct handlers.

  `event-store` is a map of:
   - a :send function. It receives an event (map with at least :event/name
   and any other keywords related to the event).
   - an :events function. It responds with the list of events in the store."
  [event-store]
  (let [{send-to-store :send stored-events :events} event-store]
    (r/ring-handler
      (r/router
        [["/tasks/added"
          {:post       (handle-task-event "task-added" send-to-store)
           :parameters {:body {:uri :task/uri :title :task/title}}}]
         ["/tasks/completed"
          {:post       (handle-task-event "task-completed" send-to-store)
           :parameters {:body {:uri :task/uri :title :task/title}}}]
         ["/tasks" {:get (tasks-snapshot-from stored-events)}]]
        {:data
         {:coercion   reitit.coercion.spec/coercion
          :middleware [mw/wrap-format
                       rrc/coerce-exceptions-middleware
                       rrc/coerce-request-middleware
                       rrc/coerce-response-middleware]}})
      (r/create-default-handler))))

(defn in-memory-event-store []
  (let [events (atom [])]
    {:send   (fn [event]
               (when (s/valid? ::event event)
                 (swap! events conj event)))
     :events (fn [] @events)}))

(defn -main [& args] (j/run-jetty (app (in-memory-event-store)) {:port 3000}))
